### ISSUE TITLE
`Self` type infer error when using overrided `__getitem__`

### ISSUE THREAD

**Issue by yanyongyu** [microsoft/pyright#6500]
**Describe the bug**

`__getitem__` should return `Self` type but get custom `Container` class instead.

**Code or Screenshots**

[playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbAA0UYAbkSCWAIYAmlNdISzRAUHwMYlGAZ2FQAwmBQxGqOgG0AKgF0AFCSTCYS5QEoAXDyjGoAAXYhO3Iye7AoAfQdoiMJDCIQnq4aQpRGEDRhfUxpXSgAWgA%2BQj9DE0SoADpUm2N0qDsoFxgfP0pUbgAPUNQYCJi4sgSk4xAiWkYSB3gEInyyeSKiYr1MxIaYAFcQFChfLp6%2BoA)

```python
from typing import Self, overload, override


class Container[T](list[T]):
    @override
    def __getitem__(self, args: int) -> Self:
        ...
    
    def get(self, index: int) -> Self:
        reveal_type(self[index])  # type: Container[T@Container]
        return self[index]  # error
```

**VS Code extension or command-line**

Pyright 1.1.336, command-line

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
